cache/network/api.go

```diff
@@ -140,24 +155,156 @@ func (c apiClient) uploadArchive(archivePath, uploadMethod, uploadURL string, he
 	if err != nil {
 		c.logger.Warnf("error while dumping response: %s", err)
 	}
-	c.logger.Debugf("Response dump: %s", string(dump))
+	c.logger.Debugf("Chunk response dump: %s", string(dump))
 
 	if resp.StatusCode != http.StatusOK {
-		return unwrapError(resp)
+		return "", unwrapError(resp)
+	}
+
+	etag := resp.Header.Get("ETag")
+	fmt.Printf("Etag %s for %s of size %v \n\r", etag, uploadURL.URL, size)
+
+	return etag, nil
+}
+
+func (c apiClient) uploadArchive(archivePath string, chunkSize, chunkCount, lastChunkSize int, uploadURLs []uploadURL) ([]string, error) {
+
+	fmt.Printf("Uploading archive: %s \n\r", archivePath)
+	fmt.Printf("Chunk size, count, last chunk size: %d, %d, %d \n\r", chunkSize, chunkCount, lastChunkSize)
+	fmt.Printf("Upload urls: %+v \n\r", uploadURLs)
+
+	file, err := os.Open(archivePath)
+	if err != nil {
+		return nil, fmt.Errorf("open file: %s", err)
+	}
+	defer file.Close()
+
+	if chunkCount == 1 {
+		fileInfo, err := os.Stat(archivePath)
+		if err != nil {
+			return nil, fmt.Errorf("stat file: %s", err)
+		}
+		c.logger.Debugf("Uploading single chunk (non multipart upload)", uploadURLs[0].URL)
+
+		_, err = c.uploadArchiveChunk(uploadURLs[0], file, fileInfo.Size())
+		if err != nil {
+			return nil, fmt.Errorf("upload single chunk (non multipart): %s", err)
+		}
+		return nil, nil
+	}
+
+	type Job struct {
+		ChunkNumber int
+		ChunkStart  int64
+		ChunkSize   int64
+		UploadURL   uploadURL
+	}
+
+	jobs := make(chan Job, chunkCount)
+	results := make(chan struct {
+		ChunkNumber int
+		ChunkSize   int64
+		ETag        string
+	}, chunkCount)
+	errors := make(chan error, chunkCount)
+	wg := sync.WaitGroup{}
+
+	worker := func() {
+		for job := range jobs {
+
+			chunkData, err := io.ReadAll(io.NewSectionReader(file, job.ChunkStart, job.ChunkSize))
+			if err != nil {
+				errors <- fmt.Errorf("read chunk %d: %s", job.ChunkNumber, err)
+				wg.Done()
+				continue
+			}
+
+			fmt.Printf("Uploading chunk %d to %s, size: %v, chunk start: %v \n\r", job.ChunkNumber, job.UploadURL.URL, job.ChunkSize, job.ChunkStart)
+			etag, err := c.uploadArchiveChunk(job.UploadURL, chunkData, int64(len(chunkData)))
+			if err != nil {
+				fmt.Printf("Error uploading chunk %d to %s \n\r", job.ChunkNumber, job.UploadURL.URL)
+				errors <- fmt.Errorf("upload chunk part %d: %s", job.ChunkNumber, err)
+				wg.Done()
+				continue
+			}
+
+			results <- struct {
+				ChunkNumber int
+				ChunkSize   int64
+				ETag        string
+			}{job.ChunkNumber, job.ChunkSize, etag}
+			wg.Done()
+		}
+	}
+
+	workerCount := 25 // Number of concurrent workers
+	for w := 0; w < workerCount; w++ {
+		go worker()
+	}
+
+	for i := 0; i < chunkCount; i++ {
+		chunkStart := int64(i) * int64(chunkSize)
+		chunkEnd := int64(chunkSize)
```

Comment:
The end of the chunk is at `chunkStart + chunkSize` or `chunkStart + lastChunkSize`. I see you're actually passing the size of it, so let's just call it `currentChunkSize` or similar